### PR TITLE
sql/pgwire: Avoid at least one allocation for most Datum types when using PG's text format

### PR DESCRIPTION
**Pull request by nvb**
The first commit avoids an allocation for:
- Each `DDecimal` Datum
- Each `DInterval` Datum
- Each element in a `DTuple`
- Each element in a `DArray`

The second commit avoids an allocation for:
- Each `DDate` Datum
- Each `DTimestamp` Datum
- Each `DTimestampTZ` Datum

---
This change is [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/10531)

### CODE REVIEW

File: pkg/sql/pgwire/encoding.go [hunk 1]
```diff
@@ -153,7 +154,12 @@ func (b *readBuffer) getUint32() (uint32, error) {
 type writeBuffer struct {
 	wrapped bytes.Buffer
 	err     error
-	putbuf  [64]byte
+
+	// These two buffers are used as temporary storage. Use putbuf when the
+	// length of the required temp space is known. Use variablePutbuf when the length
+	// of the required temp space is unknown, or when a bytes.Buffer is needed.
+	putbuf         [64]byte
+	variablePutbuf bytes.Buffer
```
Comment (nvanbenschoten) [inline pkg/sql/pgwire/encoding.go]: I'm all ears if you have a more elegant solution to this, keeping in mind `Datum.Format`'s signature.

**Comment by tamird** [comment]
What do the benchmarks say?

:lgtm: mod the comment about replacing `putbuf`

---

Reviewed 3 of 3 files at r3, 2 of 2 files at r4.
Review status: all files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

*[pkg/sql/pgwire/encoding.go, line 162 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10531#-KW1K5B8w4VJAaFVk0Zz-r2-162:-KW2X5LhmAg-1OSxF4XO:b-i14qwt) ([raw file](https://github.com/cockroachdb/cockroach/blob/08a7f932294b57b5fa9c80ffa5b7e8bef7438167/pkg/sql/pgwire/encoding.go#L162)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary>
> I'm all ears if you have a more elegant solution to this, keeping in mind `Datum.Format`'s signature.
</details>
Looks OK to me, but given this is where you're headed, perhaps you could fully replace `putbuf` with this variable-sized buffer?

It is used as an argument to PutUint{16,32,64}; those calls can be replaced with `binary.Write`. Trickier would be replacing the `Append{Int,Float}` calls, though perhaps `fmt.Fprtinf` could be used? The last use is `initMsg` which sets the first byte to `0` before calling `wrapped.Write`, but there is no value in saving a call to `Write` here, so that should be easy to replace, too.

---

*[pkg/sql/pgwire/types.go, line 188 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10531#-KW2YUFxZmizxJ9DdI7q:-KW2YUFybsUpBQEwXHph:bst9ack) ([raw file](https://github.com/cockroachdb/cockroach/blob/407fd675e8ae95f68283f26ede2fc4da856af0e6/pkg/sql/pgwire/types.go#L188)):*
> ```Go
> 
> 	case *parser.DTuple:
> 		tb := &b.variablePutbuf
> ```

just remove this local binding?

---

*[pkg/sql/pgwire/types.go, line 204 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10531#-KW2YYsfOC8UK6hEpS3n:-KW2YYsfOC8UK6hEpS3o:b-7ropln) ([raw file](https://github.com/cockroachdb/cockroach/blob/407fd675e8ae95f68283f26ede2fc4da856af0e6/pkg/sql/pgwire/types.go#L204)):*
> ```Go
> 
> 	case *parser.DArray:
> 		tb := &b.variablePutbuf
> ```

ditto

---

*[pkg/sql/pgwire/types.go, line 340 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/10531#-KW2Z5D3ac1RO6C2IPnu:-KW2Z5D3ac1RO6C2IPnv:bfqrg5c) ([raw file](https://github.com/cockroachdb/cockroach/blob/b4e5431483ad979916690efe34c8f42e97b20c3a/pkg/sql/pgwire/types.go#L340)):*
> ```Go
> 
> // formatTs formats t with an optional offset into a format lib/pq understands,
> // appending to the provided tmp buffer if it is large enough and returning the
> ```

s/if it is large enough/ - it always calls `append` so it will reallocate if needed.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10531#-:-KW2YbmDkVp1QsW2QgdX:bbra9f3)*

**Comment by nvanbenschoten** [comment]
Unfortunately, none of our current benchmarks stress this.

---

Review status: 1 of 3 files reviewed at latest revision, 4 unresolved discussions.

---

*[pkg/sql/pgwire/encoding.go, line 162 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10531#-KW1K5B8w4VJAaFVk0Zz-r2-162:-KWYvl9dwDQglIuuiEGE:b-76u6ng) ([raw file](https://github.com/cockroachdb/cockroach/blob/08a7f932294b57b5fa9c80ffa5b7e8bef7438167/pkg/sql/pgwire/encoding.go#L162)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> Looks OK to me, but given this is where you're headed, perhaps you could fully replace `putbuf` with this variable-sized buffer?
> 
> It is used as an argument to PutUint{16,32,64}; those calls can be replaced with `binary.Write`. Trickier would be replacing the `Append{Int,Float}` calls, though perhaps `fmt.Fprtinf` could be used? The last use is `initMsg` which sets the first byte to `0` before calling `wrapped.Write`, but there is no value in saving a call to `Write` here, so that should be easy to replace, too.
</details>
I initially considered that exact approach, but it seems like extra overhead and effort in all cases for no real benefit.

---

*[pkg/sql/pgwire/types.go, line 188 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10531#-KW2YUFxZmizxJ9DdI7q:-KWYwDPkNS3ltj_NHcYy:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/407fd675e8ae95f68283f26ede2fc4da856af0e6/pkg/sql/pgwire/types.go#L188)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> just remove this local binding?
</details>
Done.

---

*[pkg/sql/pgwire/types.go, line 204 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10531#-KW2YYsfOC8UK6hEpS3n:-KWYwDzApe0dcvigOXQh:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/407fd675e8ae95f68283f26ede2fc4da856af0e6/pkg/sql/pgwire/types.go#L204)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> ditto
</details>
Done.

---

*[pkg/sql/pgwire/types.go, line 340 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/10531#-KW2Z5D3ac1RO6C2IPnu:-KWYx5ehM4ko8K1aKdnI:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b4e5431483ad979916690efe34c8f42e97b20c3a/pkg/sql/pgwire/types.go#L340)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> s/if it is large enough/ - it always calls `append` so it will reallocate if needed.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10531#-:-KWYtxL5i8v-qxLJ9srt:b-vgoce7)*

**Comment by tamird** [comment]
Reviewed 1 of 3 files at r3, 2 of 2 files at r5, 2 of 2 files at r6.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

*[pkg/sql/pgwire/encoding.go, line 162 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10531#-KW1K5B8w4VJAaFVk0Zz-r2-162:-KWYzfbGo9ZWqI3Zdpkp:b8gjbvg) ([raw file](https://github.com/cockroachdb/cockroach/blob/08a7f932294b57b5fa9c80ffa5b7e8bef7438167/pkg/sql/pgwire/encoding.go#L162)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary>
> I initially considered that exact approach, but it seems like extra overhead and effort in all cases for no real benefit.
</details>
What overhead? The benefit is just less random crap in the writeBuffer structure.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10531)*

**Comment by nvanbenschoten** [comment]
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[pkg/sql/pgwire/encoding.go, line 162 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10531#-KW1K5B8w4VJAaFVk0Zz-r2-162:-KWdzoZnnbb5QU-UzbOE:bx2ummf) ([raw file](https://github.com/cockroachdb/cockroach/blob/08a7f932294b57b5fa9c80ffa5b7e8bef7438167/pkg/sql/pgwire/encoding.go#L162)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> What overhead? The benefit is just less random crap in the writeBuffer structure.
</details>
Compare `binary.BigEndian.PutUint16` to `binary.Write` and `strconv.AppendInt` to `fmt.Fprintf`. Both alternatives are much larger functions that do a lot more work and usher parameters through interfaces. This was meant as an optimization, so I don't want to risk losing performance.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10531)*

**Comment by tamird** [comment]
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[pkg/sql/pgwire/encoding.go, line 162 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10531#-KW1K5B8w4VJAaFVk0Zz-r2-162:-KWe9orb1B9cABef4WRF:b43usvw) ([raw file](https://github.com/cockroachdb/cockroach/blob/08a7f932294b57b5fa9c80ffa5b7e8bef7438167/pkg/sql/pgwire/encoding.go#L162)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary>
> Compare `binary.BigEndian.PutUint16` to `binary.Write` and `strconv.AppendInt` to `fmt.Fprintf`. Both alternatives are much larger functions that do a lot more work and usher parameters through interfaces. This was meant as an optimization, so I don't want to risk losing performance.
</details>
Please add a comment, then. FWIW, this is why (in my opinion) it's not a good idea to optimize without benchmarks, so I'd really prefer that you write a benchmark here. Remember https://github.com/cockroachdb/cockroach/pull/6661? :trollface: 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10531)*

**Comment by nvanbenschoten** [comment]
Review status: 0 of 3 files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

*[pkg/sql/pgwire/encoding.go, line 162 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10531#-KW1K5B8w4VJAaFVk0Zz-r2-162:-KWejoZuHK55YNLrwQCP:b-f5dg6d) ([raw file](https://github.com/cockroachdb/cockroach/blob/08a7f932294b57b5fa9c80ffa5b7e8bef7438167/pkg/sql/pgwire/encoding.go#L162)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> Please add a comment, then. FWIW, this is why (in my opinion) it's not a good idea to optimize without benchmarks, so I'd really prefer that you write a benchmark here. Remember https://github.com/cockroachdb/cockroach/pull/6661? :trollface: 
</details>
Ok, I've added a first commit full of benchmarks. The results are in the commit messages. Improvements were as expected. The result of attempting to use `fmt.Fprintf` was:

```
name                 old time/op    new time/op    delta
WriteTextInt-4         74.8ns ± 3%   199.9ns ± 2%  +167.32%        (p=0.000 n=10+10)
WriteTextFloat-4        205ns ± 2%     535ns ± 4%  +160.65%        (p=0.000 n=10+10)

name                 old alloc/op   new alloc/op   delta
WriteTextInt-4         0.00B ±NaN%     8.00B ± 0%     +Inf%        (p=0.000 n=10+10)
WriteTextFloat-4       0.00B ±NaN%     8.00B ± 0%     +Inf%        (p=0.000 n=10+10)

name                 old allocs/op  new allocs/op  delta
WriteTextInt-4          0.00 ±NaN%      1.00 ± 0%     +Inf%        (p=0.000 n=10+10)
WriteTextFloat-4        0.00 ±NaN%      1.00 ± 0%     +Inf%        (p=0.000 n=10+10)
```

This is a clear showstopper for that approach.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10531)*

**Comment by tamird** [comment]
:lgtm:

---

Reviewed 1 of 2 files at r8, 2 of 2 files at r9.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[pkg/sql/pgwire/encoding.go, line 162 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10531#-KW1K5B8w4VJAaFVk0Zz-r2-162:-KWf9Job8wasEPrWPMnC:bwa7mb8) ([raw file](https://github.com/cockroachdb/cockroach/blob/08a7f932294b57b5fa9c80ffa5b7e8bef7438167/pkg/sql/pgwire/encoding.go#L162)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary>
> Ok, I've added a first commit full of benchmarks. The results are in the commit messages. Improvements were as expected. The result of attempting to use `fmt.Fprintf` was:
> 
> ```
> name                 old time/op    new time/op    delta
> WriteTextInt-4         74.8ns ± 3%   199.9ns ± 2%  +167.32%        (p=0.000 n=10+10)
> WriteTextFloat-4        205ns ± 2%     535ns ± 4%  +160.65%        (p=0.000 n=10+10)
> 
> name                 old alloc/op   new alloc/op   delta
> WriteTextInt-4         0.00B ±NaN%     8.00B ± 0%     +Inf%        (p=0.000 n=10+10)
> WriteTextFloat-4       0.00B ±NaN%     8.00B ± 0%     +Inf%        (p=0.000 n=10+10)
> 
> name                 old allocs/op  new allocs/op  delta
> WriteTextInt-4          0.00 ±NaN%      1.00 ± 0%     +Inf%        (p=0.000 n=10+10)
> WriteTextFloat-4        0.00 ±NaN%      1.00 ± 0%     +Inf%        (p=0.000 n=10+10)
> ```
> 
> This is a clear showstopper for that approach.
</details>
Cool, thanks for doing that! Could you leave a comment, though?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10531#-:-KWf9Q0hyeW1VgZOnsjn:bnfp4nl)*

**Comment by nvanbenschoten** [comment]
TFTR!

---

Review status: 2 of 3 files reviewed at latest revision, 1 unresolved discussion.

---

*[pkg/sql/pgwire/encoding.go, line 162 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10531#-KW1K5B8w4VJAaFVk0Zz-r2-162:-KWfPVrZSH_CYK5wcd3H:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/08a7f932294b57b5fa9c80ffa5b7e8bef7438167/pkg/sql/pgwire/encoding.go#L162)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> Cool, thanks for doing that! Could you leave a comment, though?
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10531#-:-KWfPWYl6MPGWhxOHyg2:b-814g8k)*

**Pull request merged**